lti_consumer/lti_1p3/extensions/rest_framework/serializers.py

```diff
@@ -227,3 +229,70 @@ class Meta:
             'resultMaximum',
             'comment',
         )
+
+
+def create_lti_nrps_member_serializer():
+    """
+    Extend UserReadOnlySerializer and return extended Serializer class
+    """
+    UserReadOnlySerializer = compat.get_user_readonly_serializer()
+
+    class ContextMemberSerializer(UserReadOnlySerializer):
+        """
+        Extend UserReadOnlySerializer to serialize User & User Profile information
+        """
+
+        def to_representation(self, user):
+            """
+            Override to_representation to add roles and user_id (external_id) accoding to
+            LTI NRPS Spec - https://www.imsglobal.org/spec/lti-nrps/v2p0#sharing-of-personal-data
+            """
+            result = super(ContextMemberSerializer, self).to_representation(user)
+
+            # prepare & set member roles
+            roles = []
+
+            if user.courseenrollment_set.count() > 0:
+                roles += [role['role'] for role in LTI_1P3_CONTEXT_ROLE_MAP['student']]
+
+            for accessrole in user.courseaccessrole_set.all():
+                if LTI_1P3_CONTEXT_ROLE_MAP.get(accessrole.role):
+                    roles += [role['role'] for role in LTI_1P3_CONTEXT_ROLE_MAP[accessrole.role]]
+
+            result['roles'] = set(roles)
+
+            # set user_id to external_user_id
+            externalid, _ = compat.get_or_create_externalid(user)
```

Comment:
Even though it will be an atomic operation, the number of queries will be large. I've opened a https://github.com/edx/edx-platform/pull/25844 for batch operation. If that gets approved we can use that here.